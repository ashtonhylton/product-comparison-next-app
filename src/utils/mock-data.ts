export const mockProducts = [
  {
    id: 1,
    title: "Product Title #1",
    price: 109.95,
    description: "Product Description #1",
    category: "Product Category #1",
    image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
    rating: {
      rate: 3.9,
      count: 120,
    },
  },
  {
    id: 2,
    title: "Product Title #2",
    price: 22.3,
    description: "Product Description #2",
    category: "Product Category #2",
    image:
      "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg",
    rating: {
      rate: 4.1,
      count: 259,
    },
  },
  {
    id: 3,
    title: "Product Title #3",
    price: 55.99,
    description: "Product Description #3",
    category: "Product Category #3",
    image: "https://fakestoreapi.com/img/71li-ujtlUL._AC_UX679_.jpg",
    rating: {
      rate: 4.7,
      count: 500,
    },
  },
  {
    id: 4,
    title: "Product Title #4",
    price: 15.99,
    description: "Product Description #4",
    category: "Product Category #4",
    image: "https://fakestoreapi.com/img/71YXzeOuslL._AC_UY879_.jpg",
    rating: {
      rate: 2.1,
      count: 430,
    },
  },
  {
    id: 5,
    title: "Product Title #5",
    price: 695,
    description: "Product Description #5",
    Category: "Product Category #5",
    image: "https://fakestoreapi.com/img/71pWzhdJNwL._AC_UL640_QL65_ML3_.jpg",
    rating: {
      rate: 4.67,
      count: 400,
    },
  },
];
