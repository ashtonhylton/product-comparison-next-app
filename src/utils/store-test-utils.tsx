import React, { PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import { Provider } from "react-redux";

import { makeStore, type AppStore, type RootState } from "../store/store";
import { productSlice } from "../store/slices/products-slice";

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: Partial<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  extendedRenderOptions: ExtendedRenderOptions = {}
) {
  const {
    preloadedState = {
      product: productSlice.getInitialState(),
    },
    // Automatically create a store instance if no store was passed in
    store = makeStore(preloadedState),
    ...renderOptions
  } = extendedRenderOptions;

  const Wrapper = ({ children }: PropsWithChildren) => (
    <Provider store={store}>{children}</Provider>
  );

  // Return an object with the store and all of RTL's query functions
  return {
    store,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
}
