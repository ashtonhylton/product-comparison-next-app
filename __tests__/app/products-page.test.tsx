import { act, render, screen } from "@testing-library/react";
import ProductsPage from "@app/(routes)/products/page";
import { renderWithProviders } from "@/src/utils/store-test-utils";
import { http, HttpResponse, delay } from "msw";
import { setupServer } from "msw/node";
import { mockProducts } from "@/src/utils/mock-data";

export const env = {
  productCompareLimit: process.env.NEXT_PUBLIC_PRODUCT_COMPARE_LIMIT,
  fakeStoreApiBaseUrl: process.env.NEXT_PUBLIC_FAKE_STORE_API_BASE_URL,
};

const handlers = [
  http.get(env.fakeStoreApiBaseUrl + "/products", async () => {
    await delay(200);
    return HttpResponse.json(mockProducts);
  }),
];

const server = setupServer(...handlers);

describe("Product Page with Products", () => {
  beforeAll(() => server?.listen());

  afterEach(() => server?.resetHandlers());

  afterAll(() => server?.close());

  test("Page Render Snapshot", async () => {
    const page = renderWithProviders(<ProductsPage />);

    expect(page).toMatchSnapshot();
  });

  it("It displays initial header and label text", () => {
    // given
    renderWithProviders(<ProductsPage />);

    // when
    const h1 = screen.getByRole("heading", { level: 1 });
    const h2 = screen.getByRole("heading", { level: 2 });
    const selectProductsPrompt = screen.getByText(
      /Select some products to diplay the comparison table here/,
      { selector: "p" }
    );
    const comparisonSectionTitle = screen.queryByText(/Products Compared/i, {
      selector: "h3",
    });

    //then
    expect(h1).toHaveTextContent("Products Page");
    expect(h2).toHaveTextContent(
      `Select Up To ${env.productCompareLimit} Products To Compare`
    );
    expect(selectProductsPrompt).toBeInTheDocument();
    expect(comparisonSectionTitle).not.toBeInTheDocument();
  });

  it("It displays mesage to user while awaiting Product API response", () => {
    // given
    renderWithProviders(<ProductsPage />);

    // when
    const loadingProductsText = screen.getByText(/Loading Products.../i);

    //then
    expect(loadingProductsText).toBeInTheDocument();
  });

  it("Renders Product Page with 5 Products", async () => {
    // given
    act(() => renderWithProviders(<ProductsPage />));

    // when
    await screen.findByText(/Products found \(5\)/i);

    //then
    expect(screen.getByText(/Product Title #1/i)).toBeInTheDocument();
    expect(screen.getByText(/Product Title #2/i)).toBeInTheDocument();
    expect(screen.getByText(/Product Title #3/i)).toBeInTheDocument();
    expect(screen.getByText(/Product Title #4/i)).toBeInTheDocument();
    expect(screen.getByText(/Product Title #5/i)).toBeInTheDocument();
  });

  it("It displays comparison section when products are selected", async () => {
    // given
    act(() => renderWithProviders(<ProductsPage />));

    // when
    await screen.findByText(/Products found \(5\)/i);

    const priceOrderButton = screen.getAllByRole("button", { name: /Select/i });
    await act(async () => await priceOrderButton[0].click());
    await act(async () => await priceOrderButton[1].click());

    //then
    const selectProductsPrompt = screen.queryByText(
      /Select some products to diplay the comparison table here/,
      { selector: "p" }
    );
    const comparisonSectionTitle = screen.queryByText(/Products Compared/i, {
      selector: "h3",
    });
    expect(comparisonSectionTitle).toBeInTheDocument();
    expect(selectProductsPrompt).not.toBeInTheDocument();
  });
});
