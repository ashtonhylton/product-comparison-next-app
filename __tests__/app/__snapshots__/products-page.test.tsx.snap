// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Page with Products Page Render Snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h1
        class="text-2xl text-center font-bold m-4"
      >
        Products Page
      </h1>
      <div
        class="text-center mb-20"
      >
        <h2
          class="font-semibold"
        >
          Select Up To 
          4
           Products To Compare
        </h2>
        <p>
          There are currently 
          0
           of
           
          4
           Products being compared currently
        </p>
        <p>
          You can view the selected products on the
           
          <a
            class="font-bold hover:text-pear-500"
            href="/product-comparison"
          >
            Compared Products
          </a>
           
          page or
           
          <a
            class="font-bold hover:text-pear-500"
            href="#comparison-section"
          >
            below
          </a>
        </p>
      </div>
      <p
        class="text-center my-40"
      >
        <strong>
          Loading Products...
        </strong>
        <br />
        Sit tight, Products are on the way
      </p>
      <div
        class="pt-16"
        id="comparison-section"
      >
        <p
          class="text-md text-center font-bold m-4 mb-10"
        >
          Select some products to diplay the comparison table here ☺️
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <h1
      class="text-2xl text-center font-bold m-4"
    >
      Products Page
    </h1>
    <div
      class="text-center mb-20"
    >
      <h2
        class="font-semibold"
      >
        Select Up To 
        4
         Products To Compare
      </h2>
      <p>
        There are currently 
        0
         of
         
        4
         Products being compared currently
      </p>
      <p>
        You can view the selected products on the
         
        <a
          class="font-bold hover:text-pear-500"
          href="/product-comparison"
        >
          Compared Products
        </a>
         
        page or
         
        <a
          class="font-bold hover:text-pear-500"
          href="#comparison-section"
        >
          below
        </a>
      </p>
    </div>
    <p
      class="text-center my-40"
    >
      <strong>
        Loading Products...
      </strong>
      <br />
      Sit tight, Products are on the way
    </p>
    <div
      class="pt-16"
      id="comparison-section"
    >
      <p
        class="text-md text-center font-bold m-4 mb-10"
      >
        Select some products to diplay the comparison table here ☺️
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
